### PR TITLE
Cdt

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

File: org.sonarlint.eclipse.cdt/src/org/sonarlint/eclipse/cdt/internal/CProjectConfigurator.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+  private static Path createJsonFile(Path projectBaseDir, String content) throws IOException {
+    Path dir = projectBaseDir.resolve(BUILD_WRAPPER_OUTPUT_DIRNAME);
+    Path jsonFilePath = dir.resolve(BUILD_WRAPPER_OUTPUT_FILENAME);
+    Files.createDirectories(dir);
+    Files.write(jsonFilePath, content.getBytes(BUILD_WRAPPER_OUTPUT_CHARSET));
+    return jsonFilePath;
+  }
+
+  private void printEntries(List<ICLanguageSettingEntry> entries) {
```
Comment (dbmeneses) [inline org.sonarlint.eclipse.cdt/src/org/sonarlint/eclipse/cdt/internal/CProjectConfigurator.java]: from here down will be removed (forgot)

File: org.sonarlint.eclipse.cdt/src/org/sonarlint/eclipse/cdt/internal/CProjectConfigurator.java [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    return projectLocation != null ? projectLocation.toFile().toPath() : ResourcesPlugin.getWorkspace().getRoot().getLocation().toFile().toPath();
+  }
+
+  private static void cleanCppCache(IProject project) {
+    SonarLintProject sonarProject = SonarLintProject.getInstance(project);
+    IPath projectSpecificWorkDir = project.getWorkingLocation(SonarLintCorePlugin.PLUGIN_ID);
+    Path cppCacheDir = projectSpecificWorkDir.toFile().toPath().resolve("cfamily").resolve("pph");
+    if (Files.exists(cppCacheDir)) {
+      SonarLintCorePlugin.getDefault().debug("Deleting cache: " + cppCacheDir.toString());
+      // FileUtils.deleteDirectory(cppCacheDir);
```
Comment (dbmeneses) [inline org.sonarlint.eclipse.cdt/src/org/sonarlint/eclipse/cdt/internal/CProjectConfigurator.java]: also forgot this

File: org.sonarlint.eclipse.cdt.feature/feature.properties [hunk 3]
```diff
@@ -0,0 +1,13 @@
+feature_name=SonarLint for Eclipse Java Configuration Helper
+provider_name=SonarSource
+
+description=\
+Integrate with Eclipse JDT to configure SonarLint analysis properties for Java projects
```
Comment (janos-ss) [inline org.sonarlint.eclipse.cdt.feature/feature.properties]: s/JDT/CDT/, s/Java/C++/ ?

File: org.sonarlint.eclipse.cdt/src/org/sonarlint/eclipse/cdt/internal/CProjectConfigurator.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+  private void printEntries(List<ICLanguageSettingEntry> entries) {
+    if (entries != null) {
+      for (ICLanguageSettingEntry e : entries) {
+        System.out.println("type: " + e.getKind() + ", name=" + e.getName() + ", value=" + e.getValue());
+      }
+    }
+  }
+
+  public void configureCProject2(ICProject cProject, Map<String, String> sonarProjectProperties) {
```
Comment (dbmeneses) [inline org.sonarlint.eclipse.cdt/src/org/sonarlint/eclipse/cdt/internal/CProjectConfigurator.java]: was my previous strategy, I forgot to delete it

File: org.sonarlint.eclipse.cdt.feature/feature.properties (hunk 3 above)
Comment (dbmeneses) [inline org.sonarlint.eclipse.cdt.feature/feature.properties]: good catch

File: pom.xml [hunk 5]
```diff
@@ -80,7 +83,7 @@
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     <tycho.version>0.26.0</tycho.version>
     <tycho-extras-version>${tycho.version}</tycho-extras-version>
-    <target.platform>e46</target.platform>
+    <target.platform>e45</target.platform>
```
Comment (janos-ss) [inline pom.xml]: I guess you'll change that back after rebase...

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AnalyzeProjectJob.java [hunk 6]
```diff
@@ -194,6 +197,17 @@ public Charset getCharset() {
...
+
+    @Override
+    public String contents() throws IOException {
+      return new String(Files.readAllBytes(filePath), getCharset());
+
+    }
+
+    @Override
+    public InputStream inputStream() throws IOException {
+      return Files.newInputStream(filePath);
```
Comment (janos-ss) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AnalyzeProjectJob.java]: Hm, that looks much better than the way I did it... (My head was full of IResource stuff...)

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AnalyzeProjectJob.java [hunk 7]
```diff
@@ -27,6 +27,7 @@
 import java.nio.charset.StandardCharsets;
 import java.nio.file.FileSystem;
 import java.nio.file.FileSystems;
+import java.nio.file.Files;
```
Comment (SonarTech) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AnalyzeProjectJob.java]: MINOR Remove this unused import 'java.nio.file.Files'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: org.sonarlint.eclipse.cdt/src/org/sonarlint/eclipse/cdt/internal/CProjectConfigurator.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+  private static Path getProjectBaseDir(IProject project) {
+    IPath projectLocation = project.getLocation();
+    // In some infrequent cases the project may be virtual and don't have physical location
+    return projectLocation != null ? projectLocation.toFile().toPath() : ResourcesPlugin.getWorkspace().getRoot().getLocation().toFile().toPath();
+  }
+
+  private static void cleanCppCache(IProject project) {
+    IPath projectSpecificWorkDir = project.getWorkingLocation(SonarLintCorePlugin.PLUGIN_ID);
+    Path cppCacheDir = projectSpecificWorkDir.toFile().toPath().resolve("cfamily").resolve("pph");
+    if (Files.exists(cppCacheDir)) {
```
Comment (SonarTech) [inline org.sonarlint.eclipse.cdt/src/org/sonarlint/eclipse/cdt/internal/CProjectConfigurator.java]: MAJOR Replace this with a call to the "toFile().exists" method [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3725)

File: org.sonarlint.eclipse.cdt/src/org/sonarlint/eclipse/cdt/internal/SonarCdtPlugin.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+import org.sonarlint.eclipse.core.AbstractPlugin;
+
+public class SonarCdtPlugin extends AbstractPlugin {
+
+  public static final String PLUGIN_ID = "org.sonarlint.eclipse.cdt"; //$NON-NLS-1$
+
+  private static SonarCdtPlugin plugin;
+
+  public SonarCdtPlugin() {
+    plugin = this;
```
Comment (SonarTech) [inline org.sonarlint.eclipse.cdt/src/org/sonarlint/eclipse/cdt/internal/SonarCdtPlugin.java]: CRITICAL Synchronize this lazy initialization of 'plugin' [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2444)

File: org.sonarlint.eclipse.cdt/src/org/sonarlint/eclipse/cdt/internal/BuildWrapperJsonFactory.java [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+      + "\"executable\":\"" + EXE_ID + "\","
+      + "\"stdout\":\"\","
+      + "\"stderr\":\"\""
+      + "}");
+  }
+
+  /**
+   * Copied from Jettison
+   */
+  public static String quote(String string) {
```
Comment (SonarTech) [inline org.sonarlint.eclipse.cdt/src/org/sonarlint/eclipse/cdt/internal/BuildWrapperJsonFactory.java]: CRITICAL The Cyclomatic Complexity of this method "quote" is 14 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: org.sonarlint.eclipse.cdt/src/org/sonarlint/eclipse/cdt/internal/BuildWrapperJsonFactory.java [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+        case '\n':
+          sb.append("\\n");
+          break;
+        case '\f':
+          sb.append("\\f");
+          break;
+        case '\r':
+          sb.append("\\r");
+          break;
+        default:
```
Comment (SonarTech) [inline org.sonarlint.eclipse.cdt/src/org/sonarlint/eclipse/cdt/internal/BuildWrapperJsonFactory.java]: MAJOR Reduce this switch case number of lines from 6 to at most 5, for example by extracting code into methods. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1151)

File: org.sonarlint.eclipse.cdt/src/org/sonarlint/eclipse/cdt/internal/BuildWrapperJsonFactory.java [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+
+  /**
+   * Copied from Jettison
+   */
+  public static String quote(String string) {
+    if (string == null || string.length() == 0) {
+      return "\"\"";
+    }
+
+    char c = 0;
```
Comment (SonarTech) [inline org.sonarlint.eclipse.cdt/src/org/sonarlint/eclipse/cdt/internal/BuildWrapperJsonFactory.java]: MAJOR Remove this useless assignment to local variable "c". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1854)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* CRITICAL 2 critical
* MAJOR 3 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Pull request merged**